### ISSUE TITLE
Does the currently pyCraft library support resource pack handling as some servers require "clients" to download and load packs to login into their servers.

### ISSUE THREAD

**Issue by MisterSoandSo** [ammaraskar/pyCraft#219]
I only realized one of the server I intended to write an afk client for has an enforced server resource pack. Does anyone know to handle the handshake with the server by giving it a false response that you've installed resource pack.

> --> 0x4E TimeUpdatePacket(world_age=20148, time_of_day=303694170000)
--> 0x19 DisconnectPacket(json_data='{"text":"","extra":[{"text":"You have to accept the server resource pack in order to play!\\n","color":"gray","bold":false,"italic":false,"underlined":false,"strikethrough":false,"obfuscated":false},{"text":"If you have the pack disabled, you must click ","color":"red","italic":false},{"text":"\\u0027edit\\u0027\\n","color":"white","italic":false},{"text":"and set ","color":"red","italic":false},{"text":"\\"Resource Pack\\" ","color":"yellow","italic":false},{"text":"to enabled.","color":"red","italic":false}]}')

**Comment by MisterSoandSo** [comment]
Let me rephrase my original question: Does the currently pyCraft library support resource pack handling as some servers require "clients" to download and load packs to login into their servers.

I took a look at this issue again and researched how [ORelio tackled this issue. ](https://github.com/ORelio/Minecraft-Console-Client/blob/master/MinecraftClient/Protocol/Handlers/Protocol18.cs)In his protocol handler, he handles the handshake by sending back the response of accepted and successfully loaded. I recently looked  up the packet that handle this issue to be to listen for resource_pack_send (packet id: [0x38](https://wiki.vg/Protocol#Resource_Pack_Send)) and respond back  resource_pack_status (unknown id but I'm thinking it just a response) with the following snipped code:

`// Ignore invalid length hash string
                        if (hash.Length != 40)
                            break; 
//Send back "accepted" and "successfully loaded" responses for plugins making use of resource pack mandatory
                        byte[] responseHeader = new byte[0];
                        if (protocolversion < MC110Version) //MC 1.10 does not include resource pack hash in responses SendPacket(PacketTypesOut.ResourcePackStatus, dataTypes.ConcatBytes(responseHeader, dataTypes.GetVarInt(3))); //Accepted pack
                        SendPacket(PacketTypesOut.ResourcePackStatus, dataTypes.ConcatBytes(responseHeader, dataTypes.GetVarInt(0))); //Successfully loaded`

My question would be how would translate this protocol into python. I'm sure this is a networking issue which is related to:
clientbound/status
serverbound/status
Or would it be part of play file?

Is this a likely route to handle this issue?  I'm trying to contribute this project and I would like some insight on this matter as you have developed this project for a while. @ammaraskar

**Comment by jj136975** [comment]
I've implemented it like this (and it works). Idk if it's the good way.
To handle the packet:
In pycraft/minecraft/networking/packets/clientbound/play/__init__.py, add ResourcePackSend in the packets list and add this at the bottom of your file.

class ResourcePackSend(Packet):
    @staticmethod
    def get_id(context):
        return 0x38 #check if it match your protocol version. this one is the lastest.
    packet_name = 'resource pack send'
    get_definition = staticmethod(lambda context: [
        {'url': String},
        {'hash': String}
    ])

To send response :
In pycraft/minecraft/networking/packets/serverbound/play/__init__.py, add ResourcePackStatusPacket in the packets list and add this at the bottom of your file.

class ResourcePackStatusPacket(Packet):
    @staticmethod
    def get_id(context):
        return 0x21 #check if the protocol match your version. this one is the lastest.    
    packet_name = "resource pask status"    
    get_definition = staticmethod(lambda context: [
        {'result': VarInt},
    ])

If I'm wrong, tell me.

**Comment by MisterSoandSo** [comment]
looks about right for your return response code for this handshake looks funny because I believe the server is looking for an: 

`SendPacket(PacketTypesOut.ResourcePackStatus, dataTypes.ConcatBytes(responseHeader, dataTypes.GetVarInt(3))); //Accepted pack

SendPacket(PacketTypesOut.ResourcePackStatus, dataTypes.ConcatBytes(responseHeader, dataTypes.GetVarInt(0))); //Successfully loaded`
                       
You would need to reply back that you accepted and successfully loaded the pack. So you would need something similar sent back  as a response.

**Comment by jj136975** [comment]
In fact, you set the answer into a function like this one
![image](https://user-images.githubusercontent.com/57088933/108437888-73c15e00-724e-11eb-972f-55c583d4a9a3.png)

**Comment by MisterSoandSo** [comment]
Sorry for the late response but function call was either ignored or was never called.

**Comment by jj136975** [comment]
If you have correctly implemented all the stuff above (packets + listenner) it should automaticaly accept ressource pack when you receive the packet. Don't forget to change the packet ID to the correct one according to your version. If you want more explainations or help, contact me on discord at jj136975#5850

**Comment by joeltokyo** [comment]
Your code is working for me @jj136975, could you create a pull request with your code for the main repo?

**Comment by jj136975** [comment]
Sorry I don't have the code anymore. You can do it yourself if you want, I will not claim anything.

**Comment by MisterSoandSo** [comment]
@jj136975 I'll probably contact you about it later this month as I didn't have time to handle this side project.

**Comment by jyooru** [comment]
I've created a pull request and co-authored you, @jj136975.

**Comment by MisterSoandSo** [comment]
I'm gonna closed this issue as resolved. Thanks for the help guys @jyooru and @jj136975, I'm currently working on a new issue if you guys are interested in getting the [1.17 library working ](https://github.com/MisterSoandSo/pyCraft/commit/6400602625b4d897655edddda7492b5a8d60b125).
